### DISCUSSION TITLE
adding root level pages to the searchindex

### DISCUSSION THREAD

**Discussion by allotmentandy** [tighten/jigsaw#716] [Q&A]
hi, i am looking to add a number of pages at the root level of a jigsaw site to the searchIndex

there isnt a collection called pages (as there is for posts)

is there an easy way to do this? i would prefer to have a collection in the config.php file which states which php and md pages to add to the searchIndex if possible. 

hope you can help and point me in the right direction.

thanks in advance

**Comment by husseinalhammad** [comment]
You can get all generated pages in an event listener and do whatever you want with them (e.g. generate a sitemap):
https://jigsaw.tighten.com/docs/event-listeners/

Is this what you're after?

**Comment by allotmentandy** [comment]
no, i want to add a number of pages to the searchIndex. 

so i would like to add index.blade.php, gardening.blade.php and itsupport.blade.php to the searchIndex.json that is created at the end of the build process

but these are the pages in the root of the source directory, so i want to make an array/collection of them and hard code the filenames to add to the searchIndex. 

Alternatively, i could create them in a sub directory and use a 301 html redirect to forward the old page.

**Comment by husseinalhammad** [comment]
How are you generating the `searchIndex.json`? Jigsaw doesn't create this by default.

Are you by any chance using one of the [starter templates](https://jigsaw.tighten.com/docs/starter-templates/)?

**Comment by allotmentandy** [comment]
yes i am using the blog template with the fuse search. you can see it in action here, https://allotmentandy.github.io/ 

thanks for your help with this issue

**Comment by husseinalhammad** [comment]
The blog starter template generates the JSON file in an event listener.

You can modify the `./listeners/GenerateIndex.php` class to include the pages you need. By default this only includes items from the posts collection:
https://github.com/tighten/jigsaw-blog-template/blob/main/listeners/GenerateIndex.php

**Comment by allotmentandy** [comment]
i tried that, but there is no collection for pages, so i wonder what i would do to create an additional collection one in the config.php and then add them in that GenerateIndex.php file. 

would it just be an array/collection of filenames?

**Comment by husseinalhammad** [comment]
Check the helper methods in the docs. You can use `getPages()`
https://jigsaw.tighten.com/docs/event-listeners/#helper-methods-in-$jigsaw

**Comment by allotmentandy** [comment]
i meant the root pages, i would like to add the contact, about and index pages as part of the search, as users expect that.

can anyone point me into how to so this? 

i have added the 'front matter yaml' to each page to provide the details for each.

**Reply by husseinalhammad** [reply]
Yes, you can use `getPages()` instead of `getCollection('posts')`. It will include more than you need, so you need to filter the returned pages:

```php
$data = collect($jigsaw->getPages()->filter(function ($page, $key) {
    // filter what you need
})->map(function ($page) use ($jigsaw) {
    return [
        'title' => $page->title,
        'categories' => $page->categories,
        'link' => rightTrimPath($jigsaw->getConfig('baseUrl')) . $page->getPath(),
        'snippet' => $page->getExcerpt(),
    ];
})->values());
```

Example: you may want to exclude files with certain file extensions and exclude any file from the `/assets` directory:

```php
<?php

namespace App\Listeners;

use Illuminate\Support\Str;
use TightenCo\Jigsaw\Jigsaw;

class GenerateIndex
{
    public function handle(Jigsaw $jigsaw)
    {
        $data = collect($jigsaw->getPages()->filter(function ($page, $key) {
            return !Str::endsWith($key, ['.ico', '.atom'])
                    && !Str::startsWith($key, ['/assets']);
        })->map(function ($page) use ($jigsaw) {
            return [
                'title' => $page->title,
                'categories' => $page->categories,
                'link' => rightTrimPath($jigsaw->getConfig('baseUrl')) . $page->getPath(),
                'snippet' => $page->getExcerpt(),
            ];
        })->values());

        file_put_contents($jigsaw->getDestinationPath() . '/index.json', json_encode($data));
    }
}
```

You may also need to modify `getExcerpt()` in `config.php` based on your needs for these pages.
